fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/SubspaceProviderByKeySpacePath.java

```diff
@@ -39,27 +41,30 @@
     private final KeySpacePath keySpacePath;
 
     @Nonnull
-    private final FDBRecordContext context;
+    private final Map<String, CompletableFuture<Subspace>> databases;
```

Comment:
Should this be a size-limited `Cache` instead of a `Map`? I'm not sure what the difference in overhead is, but if someone reuses the same `SubspaceProviderByKeySpacePath` for many databases, this could theoretically grow without bound. (`Cache`s are also thread safe, unlike `HashMap`s, which is the implementation used in the constructor.)

In theory, it could be an `AsyncLoadingCache`, though we probably want to add a size limit to that cache and (in this instance) probably want to eschew a time limit. I think I might want a size limit any way (because of an another change). I think that would resolve @scgray's concern about caching an exceptional result (um, almost).

Reply:
The size of this particular cache would be driven by the number of available databases, so I don't really have much concern about the size.  But, yeah, the `AsyncLoadingCache` isn't a bad idea.

Reply:
This is simply using the same cache type as FDBDatabaseFactory (see line 98) 
 private final Map<String, FDBDatabase> databases = new HashMap<>();

Reply:
Yep, @dsimmen makes a good point. I think that we can hopefully replace much of this with the `ResolvedKeySpacePath`, I think the logic as it stands now is ok.

Reply:
Well, I'm concerned about that `FDBDatabaseFactory`'s cache as well then. I think I can buy that the size shouldn't grow too big as the number of cluster files should be relatively small, but I still have concerns about thread safety given that I don't see why there wouldn't be multiple accessors to the shared cache.

Reply:
All accesses to `FDBDatabaseFactory.databases` are `synchronized`.

Reply:
Changed to use ConcurrentHashMap.